plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/service/CodeWhispererService.kt

```diff
@@ -183,34 +198,38 @@ class CodeWhispererService(private val cs: CoroutineScope) : Disposable {
         }
 
         LOG.debug {
-            "Calling CodeWhisperer service, trigger type: ${triggerTypeInfo.triggerType}" +
+            "Calling CodeWhisperer service, jobId: $currentJobId, trigger type: ${triggerTypeInfo.triggerType}" +
                 if (triggerTypeInfo.triggerType == CodewhispererTriggerType.AutoTrigger) {
                     ", auto-trigger type: ${triggerTypeInfo.automatedTriggerType}"
                 } else {
                     ""
                 }
         }
 
-        val invocationStatus = CodeWhispererInvocationStatus.getInstance()
-        if (invocationStatus.checkExistingInvocationAndSet()) {
-            return
-        }
+        CodeWhispererInvocationStatus.getInstance().startInvocation()
 
-        invokeCodeWhispererInBackground(requestContext)
+        invokeCodeWhispererInBackground(requestContext, currentJobId, latencyContext)
     }
 
-    internal fun invokeCodeWhispererInBackground(requestContext: RequestContext): Job {
-        val popup = CodeWhispererPopupManager.getInstance().initPopup()
-        Disposer.register(popup) { CodeWhispererInvocationStatus.getInstance().finishInvocation() }
+    internal fun invokeCodeWhispererInBackground(requestContext: RequestContext, currentJobId: Int, latencyContext: LatencyContext): Job? {
+        ongoingRequestsContext[currentJobId] = requestContext
+        val sessionContext = sessionContext ?: SessionContext(requestContext.project, requestContext.editor, latencyContext = latencyContext)
+
+        // In rare cases when there's an ongoing session and subsequent triggers are from a different project or editor --
+        // we will cancel the existing session(since we've already moved to a different project or editor simply return.
+        if (requestContext.project != sessionContext.project || requestContext.editor != sessionContext.editor) {
+            disposeDisplaySession(false)
+            return null
+        }
+        this.sessionContext = sessionContext
 
         val workerContexts = mutableListOf<WorkerContext>()
         // When popup is disposed we will cancel this coroutine. The only places popup can get disposed should be
         // from CodeWhispererPopupManager.cancelPopup() and CodeWhispererPopupManager.closePopup().
         // It's possible and ok that coroutine will keep running until the next time we check it's state.
         // As long as we don't show to the user extra info we are good.
-        val coroutineScope = disposableCoroutineScope(popup)
+        val coroutineScope = projectCoroutineScope(requestContext.project)
```

Comment:
`CodeWhispererService` has a service-level scope that should be used instead

Reply:
fixed